### PR TITLE
Re-enable F16 Support for Images

### PR DESCRIPTION
**Pull request by jleibs**
### What
Introduce a new arrow attribute: "attr.rerun.override_type" that allows you to override short -> half.

This overrides the type that would otherwise be extracted from FbsBaseType.

Reinstate all the F16 code-paths, add roundtrip tests, fuzzers, etc.

This does not add ergonomic f16 support for C++.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3356) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3356)
- [Docs preview](https://rerun.io/preview/ddb6eac02405c30363d88e251f5f8edd17cb8c6b/docs) 
- [Examples preview](https://rerun.io/preview/ddb6eac02405c30363d88e251f5f8edd17cb8c6b/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -102,7 +102,7 @@ enumset = "1.0.12"
 epaint = "0.22.0"
 glam = "0.22"
 gltf = "1.1"
-half = "2.3.1"
```
Comment (teh-cmc) [inline Cargo.toml:105]: This forces that feature on every crate in the workspace that uses `bytemuck`, with no way to opt-out.
Can this be specified just for the one or two crates that require it instead?

File: crates/re_types/src/archetypes/depth_image_ext.rs [hunk 2]
```diff
@@ -12,7 +12,10 @@ impl DepthImage {
     /// for treating as an image.
     ///
     /// This is useful for constructing an [`DepthImage`] from an ndarray.
-    pub fn try_from<T: TryInto<TensorData>>(data: T) -> Result<Self, ImageConstructionError<T>> {
+    pub fn try_from<T: TryInto<TensorData>>(data: T) -> Result<Self, ImageConstructionError<T>>
+    where
+        <T as TryInto<TensorData>>::Error: std::fmt::Debug,
```
Comment (teh-cmc) [inline crates/re_types/src/archetypes/depth_image_ext.rs:17]: leftover from debugging?

File: crates/re_types/src/archetypes/image_ext.rs [hunk 3]
```diff
@@ -12,7 +12,10 @@ impl Image {
     /// for treating as an image.
     ///
     /// This is useful for constructing an [`Image`] from an ndarray.
-    pub fn try_from<T: TryInto<TensorData>>(data: T) -> Result<Self, ImageConstructionError<T>> {
+    pub fn try_from<T: TryInto<TensorData>>(data: T) -> Result<Self, ImageConstructionError<T>>
+    where
+        <T as TryInto<TensorData>>::Error: std::fmt::Debug,
```
Comment (teh-cmc) [inline crates/re_types/src/archetypes/image_ext.rs:17]: ditto

File: crates/re_types/src/archetypes/segmentation_image_ext.rs [hunk 4]
```diff
@@ -12,7 +12,10 @@ impl SegmentationImage {
     /// for treating as an image.
     ///
     /// This is useful for constructing an [`SegmentationImage`] from an ndarray.
-    pub fn try_from<T: TryInto<TensorData>>(data: T) -> Result<Self, ImageConstructionError<T>> {
+    pub fn try_from<T: TryInto<TensorData>>(data: T) -> Result<Self, ImageConstructionError<T>>
+    where
+        <T as TryInto<TensorData>>::Error: std::fmt::Debug,
```
Comment (teh-cmc) [inline crates/re_types/src/archetypes/segmentation_image_ext.rs:17]: ditto

File: crates/re_viewer/data/examples_manifest.json
Comment (teh-cmc) [inline crates/re_viewer/data/examples_manifest.json:1]: What's happening here?

File: crates/re_types/definitions/rerun/attributes.fbs [hunk 5]
```diff
@@ -17,3 +17,10 @@ attribute "attr.rerun.component_recommended";
 ///
 /// Only applies to the fields of an archetype.
 attribute "attr.rerun.component_optional";
+
+/// Override the type of a field.
+///
+/// The only permitted value is "half", which can only be used to override the type of a ushort.
```
Comment (emilk) [inline crates/re_types/definitions/rerun/attributes.fbs:23]: is "f16" or "float16" perhaps more clear than "half"?

File: tests/rust/roundtrips/image/Cargo.toml [hunk 6]
```diff
@@ -12,3 +12,5 @@ rerun = { path = "../../../../crates/rerun", features = ["native_viewer"] }
 anyhow = "1.0"
 clap = { version = "4.0", features = ["derive"] }
 image.workspace = true
+ndarray.workspace = true
+half.workspace = true
```
Comment (teh-cmc) [inline tests/rust/roundtrips/image/Cargo.toml:16]: sort

Comment (teh-cmc) [inline tests/rust/roundtrips/image/Cargo.toml:16]: sort

File: crates/re_types/src/datatypes/tensor_data_ext.rs [hunk 7]
```diff
@@ -324,22 +320,28 @@ impl<D: ::ndarray::Dimension> TryFrom<::ndarray::Array<half::f16, D>> for Tensor
...
+                shape,
+                buffer: TensorBuffer::F16(
+                    value
+                        .iter()
+                        .map(|f| arrow2::types::f16::from_bits(f.to_bits()))
+                        .collect::<Vec<_>>()
+                        .into(),
+                ),
+            }),
+        }
```
Comment (teh-cmc) [inline crates/re_types/src/datatypes/tensor_data_ext.rs:342]: a `match { true, false }` is a bit weird I have to say :sweat_smile:

File: crates/re_types/src/image.rs [hunk 8]
```diff
@@ -3,8 +3,11 @@ use smallvec::{smallvec, SmallVec};
 use crate::datatypes::{TensorData, TensorDimension};
 
 #[derive(thiserror::Error, Clone, Debug)]
-pub enum ImageConstructionError<T: TryInto<TensorData>> {
-    #[error("Could not convert source to TensorData")]
+pub enum ImageConstructionError<T: TryInto<TensorData>>
+where
+    T::Error: std::fmt::Debug,
+{
+    #[error("Could not convert source to TensorData: {0:?}")]
```
Comment (teh-cmc) [inline crates/re_types/src/image.rs:10]: Why `Debug`? I'd expect `T::Error` to implement `std::error::Error` which would imply `std::fmt::Display`?

File: crates/re_types/tests/fuzzy.rs [hunk 9]
```diff
@@ -306,7 +313,13 @@ fn roundtrip() {
         // NOTE: Keep those around please, very useful when debugging.
         // eprintln!("field = {field:#?}");
         // eprintln!("array = {array:#?}");
-        eprintln!("{} = {array:#?}", field.name);
+        if field.name == "fuzz1021" {
+            // TODO(jleibs): Fields that contain Float16 apparently don't supported fmt
+            // https://github.com/jorgecarleitao/arrow2/blob/main/src/array/primitive/fmt.rs#L35
+            eprintln!("{} = Can't be printed", field.name);
```
Comment (teh-cmc) [inline crates/re_types/tests/fuzzy.rs:319]: Some context for the poor soul who sees this in their terminal:
```suggestion
            eprintln!("{} = Can't be printed (f16 not supported)", field.name);

File: crates/re_types_builder/src/objects.rs [hunk 10]
```diff
@@ -43,9 +43,13 @@ impl Objects {
 
         let include_dir_path = include_dir_path.as_ref();
 
+        // TODO(jleibs): Do we need real attributes here?
+        let attrs = Attributes::default();
+
```
Comment (teh-cmc) [inline crates/re_types_builder/src/objects.rs:48]: not sure why you'd build this all the way here..?

**Review by teh-cmc — APPROVED** [review]
:+1:

**Comment by teh-cmc** [comment]
(No idea why half of my review comments ended up outside of the review, sorry)

File: crates/re_viewer/data/examples_manifest.json
Comment (teh-cmc) [reply crates/re_viewer/data/examples_manifest.json:1]: https://github.com/rerun-io/rerun/pull/3359 is what's happening

File: crates/re_types/src/archetypes/depth_image_ext.rs (hunk 2 above)
Comment (jleibs) [reply crates/re_types/src/archetypes/depth_image_ext.rs:17]: Required for:
```
pub enum ImageConstructionError<T: TryInto<TensorData>>
where
    T::Error: std::fmt::Debug,
{
    #[error("Could not convert source to TensorData: {0:?}")]
    TensorDataConversion(T::Error),
```
to be able to work.

I could switch to `{0}` and `std::fmt::Display` but assumed Debug was more likely to be implemented.

File: crates/re_types/definitions/rerun/attributes.fbs (hunk 5 above)
Comment (jleibs) [reply crates/re_types/definitions/rerun/attributes.fbs:23]: "half" matches the terms that the way that fbs files use for their types: int, short, long, float, double.  That doesn't necessarily make it any more clear though I'm happy to use float16 instead.

File: crates/re_types/src/datatypes/tensor_data_ext.rs (hunk 7 above)
Comment (jleibs) [reply crates/re_types/src/datatypes/tensor_data_ext.rs:342]: :laughing: fair

File: crates/re_types/src/image.rs (hunk 8 above)
Comment (jleibs) [reply crates/re_types/src/image.rs:10]: Ahh, let me see if I can make that bound work.

File: crates/re_types_builder/src/objects.rs (hunk 10 above)
Comment (jleibs) [reply crates/re_types_builder/src/objects.rs:48]: Ahh, I see one gets built inside of from_raw_enum.  Gone.

**Pull request merged**